engine/baml-lib/schema-ast/src/parser/parse_type_builder_block.rs

```diff
@@ -32,54 +32,118 @@ pub(crate) fn parse_type_builder_block(
+                                    &format!("Unexpected keyword '{nested}' in dynamic type definition. Use 'dynamic class' or 'dynamic enum'."),
+                                    diagnostics.span(nested.as_span()),
+                                ));
+                            }
+                        }
 
                         Rule::type_expression_block => {
-                            let type_expr = parse_type_expression_block(
+                            let mut type_expr = parse_type_expression_block(
                                 nested,
                                 pending_block_comment.take(),
                                 diagnostics,
                             );
 
+                            // Include the dynamic keyword in the span.
+                            type_expr.span = dyn_type_expr_span.to_owned();
+
+                            // TODO: #1343 Temporary solution until we implement scoping in the AST.
+                            // We know it's dynamic. The Dynamic subtype will be
+                            // removed later because it's not supported in the
+                            // AST but we store this information here.
+                            type_expr.is_dynamic_type_def = true;
+
                             match type_expr.sub_type {
-                                SubType::Class => entries.push(TypeBuilderEntry::Class(type_expr)),
-                                SubType::Enum => entries.push(TypeBuilderEntry::Enum(type_expr)),
-                                SubType::Dynamic => {
+                                SubType::Class | SubType::Enum => {
                                     entries.push(TypeBuilderEntry::Dynamic(type_expr))
                                 }
                                 _ => {} // may need to save other somehow for error propagation
                             }
                         }
 
-                        Rule::type_alias => {
-                            let assignment = parse_assignment(nested, diagnostics);
-                            entries.push(TypeBuilderEntry::TypeAlias(assignment));
-                        }
+                        _ => parsing_catch_all(nested, "dynamic_type_expression_block"),
+                    }
+                }
+            }
 
-                        Rule::BLOCK_LEVEL_CATCH_ALL => {
-                            diagnostics.push_error(DatamodelError::new_validation_error(
-                                "Syntax error in type builder block",
-                                diagnostics.span(nested.as_span()),
-                            ))
-                        }
+            Rule::type_expression_block => {
+                let type_expr =
+                    parse_type_expression_block(current, pending_block_comment.take(), diagnostics);
 
-                        _ => parsing_catch_all(nested, "type_builder_contents"),
-                    }
+                match type_expr.sub_type {
```

Comment:
Two comments:
1. Do we need to add a case for type aliases here?
2. Is it possible to remove the wildcard match, and enumerate all the variants explicitly? Or would there be a trillion variants?